src/_pytest/python_api.py

```diff
@@ -955,117 +968,40 @@ def raises(
-            raise TypeError(msg.format(not_a))
-        else:
-            return exc
-
-    expected_exceptions = tuple(validate_exc(exc) for exc in expected_exceptions)
-
-    message = f"DID NOT RAISE {expected_exception}"
-
-    if not args:
-        match: str | re.Pattern[str] | None = kwargs.pop("match", None)
-        if kwargs:
-            msg = "Unexpected keyword arguments passed to pytest.raises: "
-            msg += ", ".join(sorted(kwargs))
-            msg += "\nUse context-manager form instead?"
-            raise TypeError(msg)
-        return RaisesContext(expected_exceptions, message, match)
-    else:
-        func = args[0]
-        if not callable(func):
-            raise TypeError(f"{func!r} object (type: {type(func)}) must be callable")
-        try:
-            func(*args[1:], **kwargs)
-        except expected_exceptions as e:  # pylint: disable=catching-non-exception
-            return _pytest._code.ExceptionInfo.from_exception(e)
-    fail(message)
-
-
-# This doesn't work with mypy for now. Use fail.Exception instead.
-raises.Exception = fail.Exception  # type: ignore
-
-
-@final
-class RaisesContext(AbstractContextManager[_pytest._code.ExceptionInfo[E]]):
-    def __init__(
-        self,
-        expected_exception: type[E] | tuple[type[E], ...],
-        message: str,
-        match_expr: str | re.Pattern[str] | None = None,
-    ) -> None:
-        self.expected_exception = expected_exception
-        self.message = message
-        self.match_expr = match_expr
-        self.excinfo: _pytest._code.ExceptionInfo[E] | None = None
-        if self.match_expr is not None:
-            re_error = None
-            try:
-                re.compile(self.match_expr)
-            except re.error as e:
-                re_error = e
-            if re_error is not None:
-                fail(f"Invalid regex pattern provided to 'match': {re_error}")
-
-    def __enter__(self) -> _pytest._code.ExceptionInfo[E]:
-        self.excinfo = _pytest._code.ExceptionInfo.for_later()
-        return self.excinfo
-
-    def __exit__(
-        self,
-        exc_type: type[BaseException] | None,
-        exc_val: BaseException | None,
-        exc_tb: TracebackType | None,
-    ) -> bool:
-        __tracebackhide__ = True
-        if exc_type is None:
-            fail(self.message)
-        assert self.excinfo is not None
-        if not issubclass(exc_type, self.expected_exception):
-            return False
-        # Cast to narrow the exception type now that it's verified.
-        exc_info = cast(tuple[type[E], E, TracebackType], (exc_type, exc_val, exc_tb))
-        self.excinfo.fill_unfilled(exc_info)
-        if self.match_expr is not None:
-            self.excinfo.match(self.match_expr)
-        return True
+    func = args[0]
+    if not callable(func):
+        raise TypeError(f"{func!r} object (type: {type(func)}) must be callable")
+    with RaisesExc(expected_exception) as excinfo:
```

Comment:
i just noted that we have a important dissonance - the raises contextmanager is reentreant, the RaisesExc one isnt

Reply:
Are you referring to a failed match now raising `AssertionError` instead of letting the original exception through unchanged (reverted in https://github.com/pytest-dev/pytest/pull/13279#discussion_r1989112717), or something else?

Reply:
I'm referring to the state of the contextmanager instances

Reply:
yes there's state - but afaik there's no relevant state that's not reset upon doing a new `__enter__` or `matches()` call. Do you have example that demonstrates the change in behavior?

Reply:
@RonnyPfannschmidt ?